File: expressionAddOperator.py
'''
Time Complexity -->
O(4^n)

Space Complexity -->
O(4^n)
'''
class Solution:
    def addOperators(self, num, target):
        def dfs(l, r, expr, cur, last, res):
            if l == r:
                if cur == target:
                    res.append(expr)
                return
            for i in range(l + 1, r + 1):
                if i == l + 1 or (i > l + 1 and num[l] != "0"): # prevent "00"
                    s, x = num[l:i], int(num[l:i])
                    if last == None:
                        dfs(i, r, s, x, x, res)
                    else:
                        dfs(i, r, expr+"+"+s, cur + x, x, res)
                        dfs(i, r, expr+"-"+s, cur - x, -x, res)
                        dfs(i, r, expr+"*"+s, cur-last+last*x, last*x, res)
        
        res = []
        dfs(0, len(num), '', 0, None, res)
        return res
           

File: Expression_Add_Operator.java
// 282. Expression Add Operators - https://leetcode.com/problems/expression-add-operators/
// Time Complexity : O(2^N) exponential
// Space Complexity : O(N) recursive stack
// Did this code successfully run on Leetcode : yes
// Any problem you faced while coding this : No

class Solution {
    List<String> res;

    public List<String> addOperators(String num, int target) {
        res = new ArrayList<>();
        if (num == null || num.length() == 0) return res;
        helper(num, target, 0, 0, 0, "");
        return res;
    }

    private void helper(String num, int target, int index, long calc, long tail, String path) {
        //base
        if (index == num.length()) {
            if (calc == target)
                res.add(path);
        }

        //logic
        for (int i = index; i < num.length(); i++) {
            //preceding zero
            if (num.charAt(index) == '0' && i != index)
                break;
            Long curr = Long.parseLong(num.substring(index, i + 1));
            if (index == 0) {
                //+
                helper(num, target, i + 1, curr, curr, path + curr);
            } else {
                //+
                helper(num, target, i + 1, calc + curr, curr, path + "+" + curr);
                //-
                helper(num, target, i + 1, calc - curr, -curr, path + "-" + curr);
                //*
                helper(num, target, i + 1, calc - tail + tail * curr, tail * curr, path + "*" + curr);
            }
        }
    }
}